Add documentation, tests  and fix the filtering logic for /Rk command

Ayushd785 opened the pull request:
**Your checklist for this pull request**

- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/dev/CONTRIBUTING.md) to this repository.

- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/dev/DEVELOPERS.md#code-style).

- [ ] I've documented every `RZ_API` function and struct this PR changes.

- [x] I've added tests that prove my changes are effective (required for changes to `RZ_API`).

- [ ] I've updated the [Rizin book](https://github.com/rizinorg/book) with the relevant information (if needed).

- [ ] I've used AI tools to generate fully or partially these code changes and I'm sure the changes are not copyrighted by somebody else.

**Detailed description**

This PR fixes the `/Rk` command (Query ROP Gadgets by constraints) which was parsing user constraints but not actually using them for filtering.

### The Bug

The `/Rk` command parsed constraints correctly via `rop_constraint_map_parse()` but then discarded them without passing them to the ROP search logic. This meant `/Rk eax=1` would return all gadgets instead of only those that set `eax` to `1`.

### Changes Made

**1. Core Implementation (`librz/include/rz_rop.h`, `librz/core/rop.c`, `librz/core/cmd/cmd_search.c`, `librz/core/cmd/cmd_search_rop.c`)**

- Added `constraints` field to `RzRopSearchContext` structure
- Pass parsed constraints from handler to search context
- Implemented `match_rop_constraint()` to check if a gadget satisfies a single constraint
- Implemented `match_constraints()` to check all constraints
- Added constraint filtering in `process_disassembly()` that analyzes gadgets and filters by user constraints

**2. Tests (`test/db/cmd/cmd_rop`)**

Added 5 new test cases:
- `/Rk eax=1` - filter gadgets that set eax to 1
- `/Rk ecx=0` - filter gadgets that set ecx to 0
- `/Rk eax=1,ecx=0` - filter with multiple constraints
- `/Rk eax=0` - verify no matches when constraint cannot be satisfied
- `/Rk invalidconstraint` - verify error handling for invalid constraints

### Screenshots- perfectly show the filtering 

**1.  /Rk eax=1
<img width="1919" height="307" alt="Screenshot From 2025-12-16 10-52-00" src="https://github.com/user-attachments/assets/1c3d1952-3147-4f64-bdbf-91018d9c5eec" />

**2. /Rk ecx=0
<img width="1918" height="738" alt="Screenshot From 2025-12-16 10-52-16" src="https://github.com/user-attachments/assets/18045f57-1738-4b3a-b729-f96e5c7cd30a" />

**3. /Rk eax=1 ecx=0
<img width="1920" height="329" alt="Screenshot From 2025-12-16 10-52-26" src="https://github.com/user-attachments/assets/0f09be69-8b3b-443a-99fe-80c82d0095d4" />




**AI Disclosure**: I used Gemini 3.0 pro for analyzing the issue and syntax fixes, GitHub Copilot for reading and understanding the codebase efficiently and deeply

**Closing issues**

Closes #5396

Ayushd785 committed:
Add documentation and test for /Rk command

wargio commented on test/db/cmd/cmd_rop:
```diff
@@ -2867,3 +2867,129 @@ Gadget size: 8
+
+NAME=/Rk query gadget with register to constant constraint
+FILE=bins/elf/analysis/x86-helloworld-phdr
+BROKEN=1
```
you add a test and is already broken? not ok.

wargio commented on test/db/cmd/cmd_rop:
```diff
@@ -2867,3 +2867,129 @@ Gadget size: 8
+
+NAME=/Rk query gadget with register to constant zero
+FILE=bins/elf/analysis/x86-helloworld-phdr
+BROKEN=1
```
same here and below

Ayushd785 replied:
i'll fix it

wargio reviewed (changes requested):
many of the new tests have the `BROKEN=1` line. remove it as it is unacceptable to have tests that are already broken.

Ayushd785 commented:
Hi, @wargio, thank you for the review!
While writing tests for /Rk, I discovered that the command parses constraints correctly but doesn't actually use them to filter the ROP gadgets.I marked the tests as BROKEN=1 thinking it would document the expected behavior for when someone implements the filtering. But I understand now that's not the right approach.

wargio commented on librz/core/cmd_descs/cmd_search.yaml:
```diff
@@ -549,6 +549,41 @@ commands:
+                comment: "Find gadgets that set a register to a constant value"
+              - text: "reg=reg"
+                comment: "Find gadgets that copy one register to another"
+              - text: "reg=reg OP const"
+                comment: "Find gadgets with register and constant arithmetic"
+              - text: "reg=reg OP reg"
```
what is OP ?

Ayushd785 replied:
OP means operation - like +, -, *, /, %, &, |, ^, <<, >>

Ayushd785 committed:
Implement constraint filtering for /Rk command

- Add constraints field to RzRopSearchContext
- Pass parsed constraints from handler to search context
- Implement match_rop_constraint() and match_constraints() in rop.c
- Filter gadgets based on user constraints in process_disassembly()
- Add tests for /Rk filtering (eax=1, ecx=0, multiple constraints)

Ayushd785 committed:
Merge branch 'dev' into fix/rk-docs-tests

Rot127 commented on librz/core/rop.c:
```diff
@@ -1504,6 +1504,60 @@ static bool match_detail_search(st64 gadget_val, RopDetailSearchCmpOp op, st64 t
+	}
+	case MOV_OP_CONST:
+	case MOV_OP_REG:
+		return true;
```
Why are those just true? Shouldn't it check if the gadget uses a register/constant at all?

Ayushd785 replied:
Yes, you are right, currently it returns true here ignoring whether the gadget actually uses the register/constant. I’ll fix this to properly validate the operation. Thanks for the feedback

Rot127 commented on test/db/cmd/cmd_rop:
```diff
@@ -2867,3 +2867,76 @@ Gadget size: 8
+/Rk eax=1
+/Rk ecx=0
+/Rk eax=1,ecx=0
+/Rk eax=0
```
```suggestion
echo "Expected no results"
/Rk eax=0
```

Rot127 reviewed (changes requested):
Just those nitpicks. Otherwise lgtm

Ayushd785 committed:
Update test/db/cmd/cmd_rop

Co-authored-by: Rot127 <45763064+Rot127@users.noreply.github.com>

Ayushd785 committed:
Fix MOV_OP_CONST and MOV_OP_REG constraint validation in /Rk

Ayushd785 committed:
Merge branch 'dev' into fix/rk-docs-tests

Ayushd785 commented:
hii @Rot127 , I’ve updated the `MOV_OP_CONST` and `MOV_OP_REG` cases to properly enforce the constraint semantics by checking that the gadget actually reads the source register(s)

Ayushd785 committed:
Fix test: Add missing echo output to expected results

Ayushd785 committed:
Add tests for MOV_OP_CONST and MOV_OP_REG constraint types

Ayushd785 commented:
i have also added tests for `MOV_OP_CONST` and `MOV_OP_REG`

Rot127 commented on librz/core/rop.c:
```diff
@@ -1504,6 +1504,73 @@ static bool match_detail_search(st64 gadget_val, RopDetailSearchCmpOp op, st64 t
+		}
+		return rz_core_rop_gadget_reg_info_has_event(gadget_info, RZ_ROP_EVENT_VAR_READ, src_reg);
+	}
+	case MOV_OP_CONST: {
+		const char *src_reg = constraint->args[SRC_REG];
+		if (!src_reg) {
+			return false;
+		}
+		return rz_core_rop_gadget_reg_info_has_event(gadget_info, RZ_ROP_EVENT_VAR_READ, src_reg);
```
Shouldn't this also check for `SRC_CONST`?

Ayushd785 committed:
Merge branch 'dev' into fix/rk-docs-tests

notxvilka reviewed (approved):
Thank you, this is quite helpful!

Ayushd785 commented:
> Thank you, this is quite helpful!


Thanks! Glad it was helpful.
